## Chapter 12 [Onlinestatsbook.com](onlinestatsbook.com) :  "Test of Means"
------  


#### Below are selected formulas and exercises from chapter 11 of the infamous onlinestatsbook.com, a highly trusted resource for learning about statistics.  

#### The formulas and exercises were chosen based on difficulty and based on if using python to understand the concept or answer the question was deemed useful.

#### Please note the below does not include the questions from the case studies.  A separate notebook for each case study can be found in this repository or is forthcoming. 

In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
%matplotlib inline

from scipy import stats

### Section 1: "Contents"

Many, if not most experiments are designed to compare means. The experiment may involve only one sample mean that is to be compared to a specific value. Or the experiment could be testing differences among many different experimental conditions, and the experimenter could be interested in comparing each mean with each of the other means. This chapter covers methods of comparing means in many different experimental situations.

### Section 2: "Testing a Single Mean"

we wish to know the probability of obtaining a sample mean of 51 **or more** when the sampling distribution of the mean has a mean of 50 and a standard deviation (standard error) of 1.667. Note this was calculated for us.  To compute this probability, we will make the assumption that the sampling distribution of the mean is normally distributed. Using the online statsbook normal calculator we can find out the probability is .274293. Here's the image:

![alt text][img1]

[img1]:http://onlinestatbook.com/2/tests_of_means/graphics/sig_mean1.gif

We can also obtain this value with python.

In [2]:
1-stats.norm(50, 1.667).cdf(51)

0.2742930981455225

The test conducted above was a one-tailed test because it computed the probability of a sample mean being one or more points higher than the hypothesized mean of 50 and the area computed was the area above 51.

To test the two-tailed hypothesis, you would compute the probability of a sample mean differing by one or more in either direction from the hypothesized mean of 50. You would do so by computing the probability of a mean being less than or equal to 49 or greater than or equal to 51.  Here's what that looks like with the onlinestatsbook normal calc:

![alt text][img1]

[img1]:http://onlinestatbook.com/2/tests_of_means/graphics/sig_mean2.gif

Here's how to calculate that in python:


In [3]:
(1-stats.norm(50, 1.667).cdf(51))+stats.norm(50, 1.667).cdf(49)

0.54858619629104499

Can also leverage the survivial function which is 1-cdf to somewhat simplify:

In [4]:
stats.norm(50, 1.667).sf(51)+ stats.norm(50, 1.667).cdf(49)

0.5485861962910451

Typically σ is not known and is estimated in a sample by s, and σM is estimated by sM. Using an example from the ADHD study, we test the null hypothesis is that the mean difference score in the population is 0 meaning the drug makes no difference.

1.  calculate the t statistic using a special case of this formula:

  ![alt text][img1]

[img1]:http://onlinestatbook.com/2/tests_of_means/graphics/t_general.gif

  ![alt text][img2]

[img2]:http://onlinestatbook.com/2/tests_of_means/graphics/t_mean.gif

where t is the value we compute for the significance test, M is the sample mean, μ is the hypothesized value of the population mean, and sM is the estimated standard error of the mean. 

2.  The mean (M) of the N = 24 difference scores is 4.958, the hypothesized value of μ is 0, and the standard deviation (s) is 7.538. obtain sM and calculate

![alt text][img3]

[img3]: http://onlinestatbook.com/2/tests_of_means/graphics/se1.gif

Therefore, t = 4.96/1.54 = 3.22.

3. Use t to calculate the shaded area.

![alt text][img4]

[img4]: http://onlinestatbook.com/2/tests_of_means/graphics/t_calc.gif


This is rather easy to do with python:

In [5]:
(1-stats.t.cdf(3.22, df = 23))*2

0.003792872940570291

##### Question 1 out of 9.
You should do the test with Z rather than t when:

answer: The numbers are sampled from a normal distribution.

##### Question 2 out of 9.
Assume you know the standard deviation of test scores is 10 and that the distribution is normal. You sample 16 scores and find that the sample mean is 25. Find the p value for a two-tailed test of the hypothesis that the population mean is 20. 


In [6]:
#need to calculate the standard error of the mean
std_err = 10/16**.5
stats.norm(25, std_err).cdf(20)*2

0.045500263896358389

##### Question 3 out of 9.
What is the standard deviation of these sample data?



Y
 -2
  1
  3
  2
 -1
  0
  4
  6


In [7]:
x = np.asarray(list(map(int,"-2 1 3 2 -1 0 4 6".split())))
x.std(ddof=1)

2.6692695630078278

##### Question 4 out of 9.
What is the estimated standard error of the mean based on these sample data? (These are the same data as the previous question.)

In [8]:
stats.sem(x)

0.94372930440884362

##### Question 5 out of 9.
What is the value of t testing the null hypothesis that the population mean is 0? (These are the same data as the previous question.)

In [9]:
t = x.mean()/stats.sem(x)
t

1.7218920641845572

##### Question 6 out of 9.
What is the two-tailed probability value testing the null hypothesis that the population mean is 0? (These are the same data as the previous question.)

In [10]:
(1-stats.t.cdf(t, df = 7))*2

0.1287617132182699

#### Question 7 out of 9.
Using these data below, what is the t statistic for a single-sample t test (null hypothesis is that μ = 0)

Y
  2.29
  2.20
  2.28
  0.65
 -0.18
 -1.18
  1.22
 -0.07

In [11]:
x = np.asarray(list(map(float,"2.29 2.20 2.28 0.65 -0.18 -1.18 1.22 -0.07".split())))
x.mean()/stats.sem(x)

1.9368538470747008

##### Question 8 out of 9.
Using these data below, what is the t statistic for a single-sample t test (null hypothesis is that μ = .5)?

In [12]:
(x.mean()-.5)/stats.sem(x)

0.8623163452302065

##### Question 9 out of 9.
Using these data below, what is the two-tailed p value for a single-sample t test (null hypothesis is that μ = .5)?

Y
  0.05
  1.80
  0.23
  1.33
  0.17
  0.48
  0.47
  0.72

In [13]:
x = np.asarray(list(map(float,"0.05 1.80 0.23 1.33 0.17 0.48 0.47 0.72".split())))
t = (x.mean()-.5)/stats.sem(x)
(1-stats.t.cdf(t, df = len(x)-1))*2

0.4932915965578788

### Section 4: "Difference between Two Means (Independent Groups)"

It is much more common for a researcher to be interested in the difference between means than in the specific values of the means themselves.  Here's a table of stats for two groups

<div class="tableHolder300">
               <table>
                <tbody><tr>
                	<th>
                    	Group
                    </th>
                	<th>
                    	n
                    </th>
                	<th>
                    	Mean
                    </th>
                	<th>
                    	Variance
                    </th>
                </tr>
                <tr>
                	<td>
                    	Females
                    </td>
                	<td>
                    	17
                    </td>
                	<td>
                    	5.353
                    </td>
                	<td>
                    	2.743
                    </td>
                </tr>
                <tr>
                	<td>
                    	Males
                    </td>
                	<td>
                    	17
                    </td>
                	<td>
                    	3.882
                    </td>
                	<td>
                    	2.985
                    </td>
                </tr>
              </tbody></table>
              </div>

In order to test whether there is a difference between population means, we are going to make three assumptions:
1.  The two populations have the same variance. This assumption is called the assumption of homogeneity of variance.
2.  The populations are normally distributed.
3.  Each value is sampled independently from each other value. This assumption requires that each subject provide only one value. If a subject provides two scores, then the scores are not independent. The analysis of data with two scores per subject is discussed later (correlated t test)

**Note**: small-to-moderate violations of assumptions 1 and 2 do not make much difference. It is important not to violate assumption 3.


As seen in the previous section:

  ![alt text][img1]

[img1]:http://onlinestatbook.com/2/tests_of_means/graphics/t_general.gif


In this case, our statistic is the difference between sample means and our hypothesized value is 0. The hypothesized value is the null hypothesis that the difference between population means is 0.



In [14]:
diff = 5.353-3.882

The next step is to compute the estimate of the standard error of the statistic. In this case, the statistic is the difference between means, so the estimated standard error of the statistic is (). Recall the formula for the standard error of the difference between means is:

  ![alt text][img1]

[img1]:http://onlinestatbook.com/2/sampling_distributions/graphics/equal_var.gif

we estimate σ2 and use that estimate in place of σ2. Since we are assuming the two population variances are the same, we estimate this variance by averaging our two sample variances. Thus, our estimate of variance is

  ![alt text][img2]

[img2]:http://onlinestatbook.com/2/estimation/graphics/MSE.gif



In [15]:
mse = (2.743 + 2.985)/2
mse

2.864

therefore the standard error would be:

In [16]:
std_err = ((2*mse)/17)**.5 #17 is the numer in each group
std_err

0.5804663439602578

and t is:

In [17]:
t = diff/std_err
t

2.534169319730126

Lastly, we compute the probability of getting a t as large or larger than 2.533 or as small or smaller than -2.533. To do this, we need to know the degrees of freedom. The degrees of freedom is the number of independent estimates of variance on which MSE is based. This is equal to (n1 - 1) + (n2 - 1)

In [18]:
df = 17*2-1

We can see below from onlinstatbook t distribution calculator that the probability value for a two-tailed test with t = t and df = df is 0.0164

  ![alt text][img1]

[img1]:http://onlinestatbook.com/2/tests_of_means/graphics/t_prob_2-tail.gif

We can also compute this fairly easily in python:

In [19]:
(1-stats.t.cdf(t,df))*2

0.016199890812905293

##### Computations for Unequal Sample Sizes (optional)

Problem with above is that MSE, the estimate of variance, counts the group with the larger sample size more than the group with the smaller sample size. 

for group_a = [3,4,5] and group_b = [2,4]:

  * M1 = 4 and M2 = 3

  * SSE = (3-4)2 + (4-4)2 + (5-4)2 + (2-3)2 + (4-3)2 = 4

Then, MSE is computed by: 

   * MSE = SSE/df

where the degrees of freedom (df) is computed as before: 

  * df = (n1 - 1) + (n2 - 1) = (3 - 1) + (2 - 1) = 3. 
  * MSE = SSE/df = 4/3 = 1.333.
  
the formula for the standard error is replaced by:

  ![alt text][img1]

[img1]:http://onlinestatbook.com/2/estimation/graphics/sed_uneq.gif

where nh is the harmonic mean of the sample sizes and is calculated as follows:

  ![alt text][img2]

[img2]:http://onlinestatbook.com/2/estimation/graphics/nh.gif   


Therefore the replacement for the stantard error is  = 1.054 and:


  * t = (4-3)/1.054 = 0.949

  * and the two-tailed p = 0.413.
  
  
##### Question 8 out of 9.
If there are 4 scores per group and the t value is 2.34, what is the p value for a two-tailed test (to 3 decimal places)? 



In [20]:
(1-stats.t.cdf(2.34,6))*2

0.057843940267015004

##### Question 9 out of 9.
What is the t for an independent-groups t test for these data? 



  G 1	  G 2
 54	 44
 39	 47
 40	 45
 45	 30
 56	 38
 51	 17
 67	 37
 48	 53
 
 note:  this data is displaying funky.  every other number is in a separate group.  

In [21]:
x = list(map(int,"54 44 39 47 40 45 45 30 56 38 51 17 67 37 48 53".split()))
x

[54, 44, 39, 47, 40, 45, 45, 30, 56, 38, 51, 17, 67, 37, 48, 53]

In [22]:
g_a = []
g_b = []
for i in range(0,len(x)):
    if i%2==0:
        g_a.append(x[i])
    else:
        g_b.append(x[i])
print(g_a)
print(g_b)

[54, 39, 40, 45, 56, 51, 67, 48]
[44, 47, 45, 30, 38, 17, 37, 53]


In [23]:
diff = np.mean(g_a)-np.mean(g_b)
mse = (np.var(g_a, ddof=1) + np.var(g_b, ddof=1))/2
std_err = ((2*mse)/len(g_a))**.5
t = diff/std_err
t

2.1619306640686724

### Section 6: "All Pairwise Comparisons Among Means"

Many experiments are designed to compare more than two conditions. An obvious way to proceed would be to do a t test of the difference between each group mean and each of the other group means. The problem with this approach is that if you did this analysis, you would have six chances to make a Type I error. The more means that are compared, the more the Type I error rate is inflated. Figure 1 shows the number of possible comparisons between pairs of means (pairwise comparisons) as a function of the number of means. 

  ![alt text][img1]

[img1]:http://onlinestatbook.com/2/tests_of_means/graphics/number_of_comparisons.gif

The figure below shows the probability of a Type I error as a function of the number of means. 

  ![alt text][img2]

[img2]:http://onlinestatbook.com/2/tests_of_means/graphics/familywise.gif


The probability of a type I error rate is high even for a small number of means.  The Type I error rate can be controlled using a test called the **Tukey Honestly Significant Difference test or Tukey HSD** for short. The Tukey HSD is based on a variation of the t distribution that takes into account the number of means being compared. This distribution is called the **studentized range distribution.**

Note:  The assumptions of the Tukey test are essentially the same as for an independent-groups t test: normality, homogeneity of variance, and independent observations. The test is quite robust to violations of normality. Violating homogeneity of variance can be more problematical than in the two-sample case since the MSE is based on data from all groups. The assumption of independence of observations is important and should not be violated.

Using the leniency study we will do an example using the data below:

<table>
  <tbody><tr>
   <th>
                        	Condition
                        </th>
                        <th> 
                        	Mean
                        </th>
                        <th> 
                        	Variance
                        </th>
                      </tr>
                      <tr> 
                        <td> 
                        	False
                        </td>
                        <td> 
                        	5.37
                        </td>
                        <td> 
                        	3.34
                        </td>
                      </tr>
                      <tr> 
                        <td> 
                        	Felt
                        </td>
                        <td> 
                        	4.91
                        </td>
                        <td> 
                        	2.83
                        </td>
                      </tr>
                      <tr> 
                        <td> 
                       		Miserable
                        </td>
                        <td> 
                        	4.91
                        </td>
                        <td> 
                        	2.11
                        </td>
                      </tr>
                      <tr> 
                        <td> 
                        	Neutral
                        </td>
                        <td> 
                        	4.12
                        </td>
                        <td> 
                        	2.32
                        </td>
                      </tr>
              </tbody></table>

1.  Compute MSE, which is simply the mean of the variances. It is equal to 2.65.

2.  Compute Q:

  ![alt text][img3]

[img3]:http://onlinestatbook.com/2/tests_of_means/graphics/ts_form.gif

   for each pair of means, where Mi is one mean, Mj is the other mean, and n is the number of scores in each group. For these data, there are 34 observations per group. The value in the denominator is 0.279.
   
   
3.  Compute p for each comparison using the Studentized Range Calculator. The degrees of freedom is equal to the total number of observations minus the number of means. For this experiment, df = 136 - 4 = 132.

<table>
               		 <tbody><tr>
                        <th>
                        	Comparison
                        </th>
                        <th> 
                        	M<sub>i</sub>-M<sub>j</sub>
                        </th>
                        <th> 
                        	Q
                        </th>
                        <th> 
                        	p
                        </th>
                      </tr>
                      <tr> 
                        <td> 
                        	False - Felt
                        </td>
                        <td> 
                        	0.46
                        </td>
                        <td> 
                        	1.65
                        </td>
                        <td> 
                        	0.649
                        </td>
                      </tr>
                      <tr> 
                        <td> 
                        	False - Miserable
                        </td>
                        <td> 
                        	0.46
                        </td>
                        <td> 
                        	1.65
                        </td>
                        <td> 
                        	0.649
                        </td>
                      </tr>
                      <tr> 
                        <td> 
                       		 False - Neutral
                        </td>
                        <td> 
                        	1.25
                        </td>
                        <td> 
                        	4.48
                        </td>
                        <td> 
                        	0.010
                        </td>
                      </tr>
                      <tr> 
                        <td> 
                        	Felt - Miserable
                        </td>
                        <td> 
                        	0.00
                        </td>
                        <td> 
                        	0.00
                        </td>
                        <td> 
                        	1.000
                        </td>
                      </tr>
                      <tr> 
                        <td> 
                        	Felt - Neutral
                        </td>
                        <td> 
                        	0.79
                        </td>
                        <td> 
                        	2.83
                        </td>
                        <td> 
                        	0.193
                        </td>
                      </tr>
                      <tr> 
                        <td> 
                        	Miserable - Neutral
                        </td>
                        <td> 
                        	0.79
                        </td>
                        <td> 
                        	2.83
                        </td>
                        <td> 
                        	0.193
                        </td>
                      </tr>
              </tbody></table>
      
The only significant comparison is between the false smile and the neutral smile.
 

Tukey for Unequal Sample Sizes (optional)
The calculation of MSE for unequal sample sizes is similar to its calculation in an independent-groups t test. Here are the steps:

1. Compute a Sum of Squares Error (SSE) using the following formula  

   ![alt text][img1]
   [img1]:http://onlinestatbook.com/2/tests_of_means/graphics/SSE.gif
    where Mi is the mean of the ith group and k is the number of groups. 

2. Compute the degrees of freedom error (dfe) by subtracting the number of groups (k) from the total number of observations (N). Therefore, dfe = N - k.

3.  Compute MSE by dividing SSE by dfe: MSE = SSE/dfe.

4.  For each comparison of means, use the harmonic mean of the n's for the two means (nh).

All other aspects of the calculations are the same as when you have equal sample sizes.

##### Question 4 out of 7.
Assume you did an experiment with 3 groups and 16 subjects per group. The sample variances in the three groups were 14, 16, and 18. The value of MSE would be


In [24]:
sum([14, 16, 18])/3

16.0

##### Question 5 out of 7.
Assume you did an experiment with 3 groups and 16 subjects per group. The sample variances in the three groups were 14, 16, and 18. Using Tukey's test to compare the means, what would be the value of Q for a comparison of the first mean (14) with the last mean (18)?


In [25]:
14-18/(16/16)**.5

-4.0

##### Question 6 out of 7.
Assume you did an experiment with 3 groups and 16 subjects per group. The sample variances in the three groups were 14, 16, and 18. Using Tukey's test to compare the means, what would be the df for the test?

In [26]:
16*3-3

45

##### Question 7 out of 7.
Assume you did an experiment with 3 groups and 16 subjects per group. The sample variances in the three groups were 14, 16, and 18. Using Tukey's test to compare the means, what would be the two-tailed probability for a comparison of the first mean (14) with the last mean (18)? 

In [27]:
from statsmodels.stats.libqsturng import psturng
mse = (14+16+18)/3
q = (18-14)/((mse/16)**.5)
n = 3
df = 45
psturng(q,n,df)

0.018714290629434527

### Section 7: "Specific Comparisons (Independent Groups)"

This section shows how to test more complex comparisons in which the comparisons among means are more complicated than simply comparing one mean with another.

The methods in this section assume that the comparison among means was decided on before looking at the data. Therefore these comparisons are called planned comparisons. A different procedure is necessary for unplanned comparisons.

**An Example:** Twelve subjects were selected from a population of high-self-esteem subjects (esteem = 1) and an additional 12 subjects were selected from a population of low-self-esteem subjects (esteem = 2). Subjects then performed on a task and (independent of how well they really did) half in each esteem category were told they succeeded (outcome = 1) and the other half were told they failed (outcome = 2). Therefore, there were six subjects in each of the four esteem/outcome combinations and 24 subjects in all. After the task, subjects were asked to rate (on a 10-point scale) how much of their outcome (success or failure) they attributed to themselves as opposed to being due to the nature of the task.  The means of the four conditions:

| Outcome       | Esteem           | Mean  |
| ------------- |:----------------:| -----:|
| Success       | High Self-Esteem | 7.333 |
|               | Low Self-Esteem  | 5.500 |
| Failure       | High Self-Esteem | 4.833 |
|               | Low Self-Esteem  | 7.833 |

We begin by asking whether, on average, subjects who were told they succeeded differed significantly from subjects who were told they failed. The means of each condition are 6.4167 and 6.3333 but how do we do a significance test for theis difference .083?

**express this difference in terms of a _linear combination_ using a set of coefficients and the means.**  So...

compute the mean of the success conditions by multiplying each success mean by 0.5 and then adding the result.  

&nbsp;&nbsp;&nbsp;(.5)(7.333) + (.5)(5.500) = the mean of the success condition 6.42.

It's just another way of computing the mean.  It follows that the difference between the two means can be expressed as:

&nbsp;&nbsp;&nbsp; .5 x 7.333 + .5 x 5.500 -(.5 x 4.833 + .5 x 7.833)

&nbsp;&nbsp;&nbsp; = L = 3.667 + 2.750 - 2.417 - 3.917 = 0.083

We therefore can compute the difference between the "success" mean and the "failure" mean by multiplying each "success" mean by 0.5, each failure mean by -0.5, and adding the results. 

Now, the question is whether our value of L is significantly different from 0. The general formula for L is:

![alt text][img1]

[img1]:http://onlinestatbook.com/2/tests_of_means/graphics/L1.gif

where ci is the ith coefficient and Mi is the ith mean. As shown above, L = 0.083. The formula for testing L for significance is shown below:

![alt text][img2]

[img2]:http://onlinestatbook.com/2/tests_of_means/graphics/L.gif

In this example:

![alt text][img3]

[img3]:http://onlinestatbook.com/2/tests_of_means/graphics/c.gif

MSE is the mean of the variances for each of the four groups.  The value of n is the number of subjects in each group. So...

![alt text][img4]

[img4]:http://onlinestatbook.com/2/tests_of_means/graphics/tl1.gif

since df = N - k...

We can use the t distribution







In [28]:
(1-stats.t.cdf(.16,20))*2

0.87448598982479497

The two tailed probability is .874 so we can say the difference between the success condition is not significant.

A more interesting question about the results is whether the effect of outcome (success or failure) differs depending on the self-esteem of the subject. For example, success may make high-self-esteem subjects more likely to attribute the outcome to themselves, whereas success may make low-self-esteem subjects less likely to attribute the outcome to themselves.

To test this, we have to test a difference between differences. Specifically, is the difference between success and failure outcomes for the high-self-esteem subjects different from the difference between success and failure outcomes for the low-self-esteem subjects? The coefficients to test this difference between differences are shown in Table:

Table 5. Coefficients for testing difference between differences.  

|Self-Esteem|Outcome|Mean  |Coeff|Product|
|-----------|-------|------|-----|-------|
|High       |Success|7.333 |1    |7.333  |
|           |Failure|4.833 |-1    |-4.833 |
|Low        |Success|5.500 |-1    |-5.500  |
|           |Failure|7.833 |1    |7.833  |

If it is hard to see where these coefficients came from, consider that our difference between differences was computed this way:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(7.33 - 4.83) - (5.50 - 7.83)

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;= 7.33 - 4.83 - 5.50 + 7.83

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;= (1)7.33 + (-1)4.83 + (-1)5.50 + (1)7.83

The values in parentheses are the coefficients.

To continue the calculations...

L = 4.83 (the difference bw the differences)

The sume of the squared coefficients is...

![alt text][img1]

[img1]:http://onlinestatbook.com/2/tests_of_means/graphics/c2.gif

And....t is equal to...

![alt text][img2]

[img2]: http://onlinestatbook.com/2/tests_of_means/graphics/tl2.gif

And computing the two tailed probability in python:
 

In [29]:
(1-stats.t.cdf(4.64,20))*2

0.00015795741947521869

The difference is highly significant.

**note**: comparisons such as this are testing what is called an interaction. In general, there is an interaction when the effect of one variable differs as a function of the level of another variable. this will be studied later.  In this example, the effect of the outcome variable is different depending on the subject's self-esteem. For the high-self-esteem subjects, success led to more self-attribution than did failure; for the low-self-esteem subjects, success led to less self-attribution than did failure.


#### Multiple Comparisons

The more comparisons you make, the greater your chance of a Type I error.  The per-comparison error rate is the probability of a Type I error for a particular comparison and is equal to alpha for each individual comparison. The familywise error rate is the probability of making one or more Type I errors in a family or set of comparisons and we can use the bonferrone inequality to approximate it.  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Familywise error rate(FW) ≤ cα


c is the number of comparisons.  

The Bonferroni inequality can be used to control the familywise error rate as follows: If you want the familywise error rate to be α, you use α/c as the per-comparison error rate. This correction, called the Bonferroni correction, will generally result in a familywise error rate less than α. Alternatively, you could multiply the probability value by c and use the original α level.


he disadvantage of controlling the familywise error rate is that it makes it more difficult to obtain a significant result for any given comparison: The more comparisons you do, the lower the per-comparison rate must be and therefore the harder it is to reach significance. That is, the power is lower when you control the familywise error rate. The advantage is that you have a lower chance of making a Type I error.

THis section has some nice examples further fleshing it out.  

#### Orthogonal Comparisons

In the preceding sections, we talked about comparisons being independent. Independent comparisons are often called orthogonal comparisons. There is a simple test to determine whether two comparisons are orthogonal: If the sum of the products of the coefficients is 0, then the comparisons are orthogonal. Here the products are orthogonal:

|Self-Esteem|Outcome|Mean  |Coeff|Product|
|-----------|-------|------|-----|-----|
|Success    |High Esteem | .5 |1    |.5   |
|           |Low Esteem | .5 |-1    |-.5  |
|Failure    |High Esteem | -.5 |-1    |.5 |
|           |Low Esteem | -.5 |1    |-.5  |

Here they are not because the sum of the products is =.5 not 0:

|Self-Esteem|Outcome|Mean  |Coeff|Product|
|-----------|-------|------|-----|-----|
|Success    |High Esteem | .5 |.5    |.25  |
|           |Low Esteem | .5 |.5    |.25  |
|Failure    |High Esteem | -.5 |0    |0 |
|           |Low Esteem | -.5 |0   |0  |





##### Question 2 out of 6.
You plan to test all pairwise comparisons among 4 means. What is the new critical p value after a Bonferroni adjustment to maintain an experiment-wise alpha of .05?


In [30]:
.05/6

0.008333333333333333

##### Question 3 out of 6.
What coefficients would be used to compare Group 1 to the average of Groups 2-4? 

Answer: 1, -.333, -.333, -.333

##### Question 4 out of 6.
In an experiment with three conditions, the means are 2, 4, and 9. What would be the value of L for the coefficients 2, -1, -1? 

In [31]:
2*2+4*-1+9*-1

-9

##### Question 5 out of 6.
Are the following two sets of coefficients orthogonal? 

-1, 0, 0, 1  
-1, 1, 0, 0 


In [32]:
-1*-1+0*1+0*0+0*1

1

##### Question 6 out of 6.
In an experiment with three conditions and 5 subjects per condition, the means are 2, 4, and 9, and the MSE is 24. 

Find t for the coefficients -2, 1, 1. 

In [33]:
L = -2*2+1*4+1*9
den = (((4+1+1)*24)/5)**.5
L/den

1.6770509831248424

### Section 9: "Difference Between Two Means (Correlated Pairs)"

What do we do when we dont have independent groups.  For example with the ADHD study, they tested the whole group under four conditions not four groups under different conditions each.


A correlated t-test is computed by first computing the difference between the two scores for each subject, calculating the mean and test whether the mean difference is significantly different from 0. 

If we used an independent t-test the difference between means would not have been found to be statistically significant. This is a typical result: correlated t tests almost always have greater power than independent-groups t tests. This is because in correlated t tests, each difference score is a comparison of performance in one condition with the performance of that same subject in another condition. This makes each subject "their own control" and keeps differences between subjects from entering into the analysis. The result is that the standard error of the difference between means is smaller in the correlated t test and, since this term is in the denominator of the formula for t, results in a larger t.

NOte: Skipped mathematics on why the stand error is lower as well as the exercised for this section.  In general for correlated pairs, the variance is the sum of the variances minus two times the product of the correlation and the standard deviation of each.

### Section 10: "Specific Comparisons (Correlated Observations)"

Note: Skipped section 10

### Section 11:  Pairwise Comparisons (Correlated Observations)

For pairwise comparisons among independent groups, the Tukey HSD test was the recommended procedure. However, when you have one group with several scores from the same subjects, the Tukey test makes an assumption that is unlikely to hold: The variance of difference scores is the same for all pairwise differences between means.

The standard practice for pairwise comparisons with correlated observations is to compare each pair of means using the method outlined in the section "Difference Between Two Means (Correlated Pairs)" with the addition of the Bonferroni correction

For example, suppose you were going to do all pairwise comparisons among four means and hold the familywise error rate at 0.05. Since there are six possible pairwise comparisons among four means, you would use 0.05/6 = 0.0083 for the per-comparison error rate.

### Exercises

##### Q1: The scores of a random sample of 8 students on a physics test are as follows: 60, 62, 67, 69, 70, 72, 75, and 78.  
a.  Test to see if the sample mean is significantly different from 65 at the .05 level. Report the t and p values.  
b.  The researcher realizes that she accidentally recorded the score that should have been 76 as 67. Are these corrected scores significantly different from 65 at the .05 level? (relevant section)

In [34]:
#a
nums = np.array([60, 62, 67, 69, 70, 72, 75, 78])
s_M = nums.std(ddof=1)/len(nums)**.5
print("t is "+str((nums.mean()-65)/s_M))
print("p is "+str((1-stats.t.cdf(1.911, df = len(nums)-1))*2))

t is 1.9111806309
p is 0.0976138663502


In [35]:
#b
nums = np.array([60, 62, 76, 69, 70, 72, 75, 78])
s_M = nums.std(ddof=1)/len(nums)**.5
print("t is "+str((nums.mean()-65)/s_M))
print("p is "+str((1-stats.t.cdf(2.293, df = len(nums)-1))*2))

t is 2.2932387102
p is 0.0555613134774


##### Q2 A (hypothetical) experiment is conducted on the effect of alcohol on perceptual motor ability. Ten subjects are each tested twice, once after having two drinks and once after having two glasses of water. The two tests were on two different days to give the alcohol a chance to wear off. Half of the subjects were given alcohol first and half were given water first. The scores of the 10 subjects are shown below. The first number for each subject is their performance in the "water" condition. Higher scores reflect better performance. Test to see if alcohol had a significant effect. Report the t and p values. (relevant section)


water
alcohol
16
13
15
13
11
10
20
18
19
17
14
11
13
10
15
15
14
11
16
16

In [36]:
nums = list(map(int,"16 13 15 13 11 10 20 18 19 17 14 11 13 10 15 15 14 11 16 16".split()))

water = []
alchohol = []

for i in range(0,len(nums)):
    if i%2==0:
        water.append(nums[i])
    else:
        alchohol.append(nums[i])
        
water = np.array(water)
alchohol = np.array(alchohol)

print(water)
print(alchohol)

[16 15 11 20 19 14 13 15 14 16]
[13 13 10 18 17 11 10 15 11 16]


In [37]:
#first the easy way
stats.ttest_rel(water,alchohol)

Ttest_relResult(statistic=5.0185701660560547, pvalue=0.00072049133854553316)

In [38]:
#now more manually to check accuracy
diff = water - alchohol
t = diff.mean()/stats.sem(diff)
print("t = " + str(t))
print("p = "+str((1-stats.t.cdf(t, df = len(diff)-1))*2))

t = 5.01857016606
p = 0.000720491338545


##### Q3 The scores on a (hypothetical) vocabulary test of a group of 20 year olds and a group of 60 year olds are shown below. 


20 yr olds
60 yr olds
27
26
26
29
21
29
24
29
15
27
18
16
17
20
12
27
13


a.  Test the mean difference for significance using the .05 level. (relevant section).  
b.  List the assumptions made in computing your answer.

In [39]:
# setup the data
nums = list(map(int,"27 26 26 29 21 29 24 29 15 27 18 16 17 20 12 27 13".split()))

younger = []
older = []

for i in range(0,len(nums)):
    if i%2==0:
        younger.append(nums[i])
    else:
        older.append(nums[i])
        
younger = np.array(younger)
older = np.array(older)

print(younger)
print(older)

[27 26 21 24 15 18 17 12 13]
[26 29 29 29 27 16 20 27]


In [40]:
#A
# calculation is done differently for diff sample sizes and diff variance.
# first we do it manually
sse = sum((younger-younger.mean())**2)+sum((older-older.mean())**2) #sum of sum of square errors both samples
df = len(younger)+len(older)-2 #degrees of freedom
mse = sse/df #mean square error
harmonic_mean = 2/(1/len(younger) +1/len(older)) #calc harmonic mean of sample sizes
std_err = ((2*mse)/harmonic_mean)**.5 #calculate the std error differently than for equal sample sizes
t = abs((younger.mean()-older.mean()))/std_err #calculate t
p = (1-stats.t.cdf(t,df))*2
print("t = " + str(t))
print("p = "+str(p))

t = 2.42364228799
p = 0.0284764860044


In [41]:
#now the easy way
t, p = stats.ttest_ind(younger,older, equal_var=False)
print("t = " + str(t))
print("p = "+str(p))

t = -2.44572278434
p = 0.0272750460293


##### Q4 The sampling distribution of a statistic is normally distributed with an estimated standard error of 12 (df = 20). (a) What is the probability that you would have gotten a mean of 107 (or more extreme) if the population parameter were 100? Is this probability significant at the .05 level (two-tailed)? (b) What is the probability that you would have gotten a mean of 95 or less (one-tailed)? Is this probability significant at the .05 level? You may want to use the t Distribution calculator for this problem.

In [42]:
#A
1-stats.norm(100,12).cdf(107)

0.27983446359970576

In [43]:
#B
stats.norm(100,12).cdf(95)

0.33846111951068963

##### Q5 How do you decide whether to use an independent groups t test or a correlated t test (test of dependent means)? 

Answer: when the groups are not independent you use a correlated t test.  i.e. within groups.

##### Q6  An experiment compared the ability of three groups of subjects to remember briefly-presented chess positions. The data are shown below. 


Non-players
Beginners
Tournament players
22.1
32.5
40.1
22.3
37.1
45.6
26.2
39.1
51.2
29.6
40.5
56.4
31.7
45.5
58.1
33.5
51.3
71.1
38.9
52.6
74.9
39.7
55.7
75.9
43.2
55.9
80.3
43.2
57.7
85.3

a. Using the Tukey HSD procedure, determine which groups are significantly different from each other at the .05 level. (relevant section)

b. Now compare each pair of groups using t-tests. Make sure to control for the familywise error rate (at 0.05) by using the Bonferroni correction. Specify the alpha level you used. 

In [44]:
#first prepare the data

nums = list(map(float,"22.1 32.5 40.1 22.3 37.1 45.6 26.2 39.1 51.2 29.6 40.5 56.4 31.7 45.5 58.1 33.5 51.3 71.1 38.9 52.6 74.9 39.7 55.7 75.9 43.2 55.9 80.3 43.2 57.7 85.3".split()))
non_players = []
beginners = []
tourney_players = []
for i in range(0,len(nums)):
    if (i+3)%3==0:
        non_players.append(nums[i])
    elif(i+3)%3==1:
        beginners.append(nums[i])
    else:
        tourney_players.append(nums[i])
        
non_players = np.array(non_players)
beginners = np.array(beginners)
tourney_players =np.array(tourney_players)

print(non_players)
print(beginners)
print(tourney_players)

[ 22.1  22.3  26.2  29.6  31.7  33.5  38.9  39.7  43.2  43.2]
[ 32.5  37.1  39.1  40.5  45.5  51.3  52.6  55.7  55.9  57.7]
[ 40.1  45.6  51.2  56.4  58.1  71.1  74.9  75.9  80.3  85.3]


In [69]:
#imported psturng above
mse = (non_players.var() + beginners.var() + tourney_players.var())/3
n = 3
df = 3*len(non_players)-3
df = 45

import itertools 
arrays = ["non_players", "beginners", "tourney_players"]
combos = list(itertools.combinations(arrays,2))
count = 0
for x,y in itertools.combinations([non_players, beginners, tourney_players],2):
    q = abs(x.mean()-y.mean())/((mse/len(x))**.5)
    print(combos[count][0]+"-"+combos[count][1] + "=  " +str(psturng(q,n,df)))
    count = count + 1

non_players-beginners=  0.0180708489014
non_players-tourney_players=  0.001
beginners-tourney_players=  0.00270618499686


In [70]:
#check with statsmodels
data = pd.DataFrame({"non_players":non_players, "beginners":beginners, "tourney_players": tourney_players})
data = data.melt()  #statsmodels requires it in a certain format
from statsmodels.stats.multicomp import pairwise_tukeyhsd
posthoc = pairwise_tukeyhsd(
    data['value'], data['variable'],
    alpha=0.05)
print(posthoc)

     Multiple Comparison of Means - Tukey HSD,FWER=0.05     
   group1        group2     meandiff  lower    upper  reject
------------------------------------------------------------
 beginners    non_players    -13.75  -26.3921 -1.1079  True 
 beginners  tourney_players   17.1    4.4579  29.7421  True 
non_players tourney_players  30.85   18.2079  43.4921  True 
------------------------------------------------------------


In [78]:
#B
from statsmodels.stats.multicomp import MultiComparison
mod = MultiComparison(data['value'], data['variable'])
rtp = mod.allpairtest(stats.ttest_ind, method="Bonferroni")
rtp[0]

<class 'statsmodels.iolib.table.SimpleTable'>